packages/react/src/components/GroupedList/GroupedListV2.base.tsx

```diff
@@ -0,0 +1,507 @@
+  const flatList = React.useRef<IGroupedItem[]>([]);
+  const isSomeGroupExpanded = React.useRef<boolean>(computeIsSomeGroupExpanded(groups));
+
+  const [version, setVersion] = React.useState({});
+  const [toggleVersion, setToggleVersion] = React.useState({});
+
+  // eslint-disable-next-line deprecation/deprecation
+  const { shouldEnterInnerZone = isInnerZoneKeystroke } = focusZoneProps;
+
+  const listView = React.useMemo(() => {
+    return flattenItems(groups, items, flatList.current, groupProps);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [groups, groupProps, items, toggleVersion, flatList]);
+
+  const getPageSpecification = React.useCallback(
+    (flattenedIndex: number): { key?: string } => {
+      const pageGroup = listView[flattenedIndex];
+      return {
+        key: pageGroup.type === 'group' ? pageGroup.group!.key : undefined,
+      };
+    },
+    [listView],
+  );
+
+  useMount(() => {
+    if (groupProps?.isAllGroupsCollapsed) {
+      setGroupsCollapsedState(groups, groupProps.isAllGroupsCollapsed);
+    }
+    events.current = new EventGroup(this);
+    if (selection) {
+      events.current.on(selection, SELECTION_CHANGE, onSelectionChange);
+    }
+  });
+
+  const onSelectionChange = React.useCallback(() => {
+    setVersion({});
+  }, [setVersion]);
+
+  useUnmount(() => {
+    events.current?.dispose();
+  });
+
+  React.useEffect(() => {
+    setVersion({});
+  }, [compact]);
+
+  React.useEffect(() => {
+    const newIsSomeGroupExpanded = computeIsSomeGroupExpanded(groups);
+    if (newIsSomeGroupExpanded !== isSomeGroupExpanded.current) {
+      isSomeGroupExpanded.current = newIsSomeGroupExpanded;
+      onGroupExpandStateChanged?.(newIsSomeGroupExpanded);
+    }
+  }, [groups, toggleVersion, onGroupExpandStateChanged, groupExpandedVersion]);
+
+  const onToggleCollapse = React.useCallback(
+    (group: IGroup): void => {
+      const onToggleCollapseFn = groupProps?.headerProps?.onToggleCollapse;
+
+      if (group) {
+        onToggleCollapseFn?.(group);
+        group.isCollapsed = !group.isCollapsed;
+        setToggleVersion({});
+        setVersion({});
+      }
+    },
+    [setToggleVersion, groupProps],
+  );
+
+  const onToggleSelectGroup = React.useCallback(
+    (group: IGroup): void => {
+      if (group && selection && selectionMode === SelectionMode.multiple) {
+        selection.toggleRangeSelected(group.startIndex, group.count);
+      }
+    },
+    [selection, selectionMode],
+  );
+
+  const onToggleSummarize = React.useCallback(
+    (group: IGroup): void => {
+      const onToggleSummarizeFn = groupProps?.showAllProps?.onToggleSummarize;
+
+      if (onToggleSummarizeFn) {
+        onToggleSummarizeFn(group);
+      } else {
+        if (group) {
+          group.isShowingAll = !group.isShowingAll;
+        }
+
+        setVersion({});
+        setToggleVersion({});
+      }
+    },
+    [groupProps],
+  );
+
+  const renderItem = React.useCallback(
```

Comment:
It would probably be best to split out an `onRender___` callback hook for each type of grouped item we'll render. Reason being that each render flow has distinct `deps`, and that would allow use to reduce re-renders caused when those change.